lib/modifiers/batch.js

```diff
@@ -0,0 +1,166 @@
+ *                                     defined, items will be written to the stream whenever
+ *                                     either one of the conditions is met.
+ * @param   {Number} [options.time]  - The interval, in milliseconds, to limit writes to.
+ *                                     During that time, all items read on the stream will
+ *                                     be collected and written as an array at the set
+ *                                     interval.
+ * @param   {Number} [options.count] - The number of items to buffer before writing all
+ *                                     of them in an array.
+ * @returns {Stream}                 - Transform stream.
+ *
+ * @example
+ * var input = through.obj();
+ *
+ * input.pipe(nodeStream.batch({ time: 100 }));
+ *
+ * input.write(1);
+ *
+ * setTimeout(function() {
+ *   input.write(2);
+ *   input.write(3);
+ * }, 100);
+ *
+ * setTimeout(function() {
+ *   input.write(4);
+ *   input.write(5);
+ * }, 200);
+ *
+ * // => [[1], [2, 3], [4, 5]]
+ * @example
+ * var input = through.obj();
+ *
+ * input.pipe(nodeStream.batch({ count: 2 }));
+ *
+ * input.write(1);
+ * input.write(2);
+ * input.write(3);
+ * input.write(4);
+ * input.write(5);
+ *
+ * // => [[1, 2], [3, 4], [5]]
+ */
+function batch(options) {
+  var opts = _.isPlainObject(options) ? options : {};
+  var milliseconds = opts.time || 0;
+  var count = opts.count || 0;
+  var cache = [];
+  var lastWrite = 0;
+  var timeout;
+
+  /**
+   * @private
+   * @description
+   * Writes all the collected data to the stream.
+   *
+   * @param {Stream} stream - The stream to write to.
+   * @returns {undefined}
+   */
+  function flushCache(stream) {
+    lastWrite = Date.now();
+
+    if (timeout) {
+      clearTimeout(timeout);
+      timeout = null;
+    }
+
+    stream.push(cache.splice(0, count || cache.length));
+  }
+
+  /**
+   * @private
+   * @description
+   * Flushes the cache using the defined timeout.
+   *
+   * @param {Stream} stream    - The stream to write to.
+   * @param {Function} onFlush - The function to call when
+                                 the cache is flushed.
+   * @returns {undefined}
+   */
+  function flushCacheTimed(stream, onFlush) {
+    var nextWrite = lastWrite + milliseconds;
+    var now = Date.now();
+
+    // we are overdue for a write, so write now
+    if (now > nextWrite) {
+      flushCache(stream);
+      onFlush();
+
+      return;
+    }
+
+    // write after the remainder between now and when
+    // the next write should be according to milliseconds
+    // and the previous write
+    timeout = setTimeout(function () {
+      flushCache(stream);
+      onFlush();
+    }, nextWrite - now);
+  }
+
+  /**
+   * @private
+   * @description
+   * Flushes the cache if it is appropriate to do so.
+   *
+   * @param {Stream} stream - The stream to write to.
+   * @returns {undefined}
+   */
+  function writeIfPossible(stream) {
+    if (!count && !milliseconds) {
```

Comment:
I tried combining this if and the next one into one, but your linting hated it and it was difficult to read when I fixed linting.

```javascript
if ( 
  // we don't have a count or a time set, 
  (!count && !milliseconds) || 
  // the count has been met 
  (count && cache.length === count) 
) { 
```

It thinks my parentheses are gratuitous, but some people hate seeing `&&` and `||` mixed together since they can't figure out what is being checked.